### DISCUSSION TITLE
Using Csla 6 Factory type in Csla 5

### DISCUSSION THREAD

**Discussion by JacoJordaan** [MarimerLLC/csla#3185] [Questions] [answered]
I want to make the upgrade from Csla 5 to Csla 6 easier by doing away with the static factory methods and instead use the factory classes with the IDataPortalFactory injected.

The idea is to put the following example class in my Csla 5 project and use this in my code. I can then gradually change all the objects. Upgrading to Csla 6 would then be much fewer coding changes once off.

Would this be possible? If so, does anyone has suggestions on how to do this?

````
public class PersonEdit
{
  public class Factory
  {
    public Factory(IDataPortalFactory factory)
    {
      Portal = factory.GetPortal<PersonEdit>();
    }

    private IDataPortal<PersonEdit> Portal { get; set; }

    public async Task<PersonEdit> GetAsync(int personId)
    {
      return await Portal.FetchAsync(personId);
    }
  }
}
````
The calling code should then change from 
`var person = await PersonEdit.GetAsync(42);` 
to something like this:
````
var dataPortalFactory = ?? //need to know if this is possible

var person = new PersonEdit.Factory(dataPortalFactory).GetAsync(42);
````

**Comment by rockfordlhotka** [comment, marked as answer]
There are a couple ways to do the calling code, depending on whether you have DI configured. I would recommend that you _do_ configure DI, and then use a service locator pattern, so that's what I'll discuss here.

First, in app startup, ensure DI is configured, and `UseCsla` is called. Also register your factory types.

```c#
  // startup code
  services.AddCsla();
  services.AddTransient<PersonEdit.Factory>();
  // ...
  // once you've created the service provider, set your static service locator property
  App.ServiceProvider = provider;
```

Then, in your `App` class (or other globally available type) implement the service locator pattern:

```c#
  public static ServiceProvider ServiceProvider { get; set; }
```

Then in your calling code where you want to use the factory:

```c#
  var factory = App.ServiceProvider.GetRequiredService<PersonEdit.Factory>();
  var person = await factory.GetAsync(personId);
```

This approach provides the easiest migration to full use of DI in the future.
